shared/js/background/newtab-tracker-stats.js

```diff
@@ -0,0 +1,348 @@
+import browser from 'webextension-polyfill'
+import constants from '../../data/constants'
+import * as browserWrapper from './wrapper.es6.js'
+import { emitter, TrackerBlockedEvent } from './before-request.es6.js'
+import tdsStorage from './storage/tds.es6'
+const { incoming, outgoing } = constants.trackerStats.events
+
+/**
+ * The extension-specific interface to tracker stats.
+ *
+ * This class handles all interaction with the extension like event handlers, alarms etc.
+ *
+ * Example:
+ *
+ * ```js
+ * const stats = new TrackerStats();
+ * const newtab = new NewTabTrackerStats(stats)
+ *
+ * await newtab.restoreFromStorage();
+ * newtab.register()
+ * ```
+ */
+export class NewTabTrackerStats {
+    /**
+     * The key to use when persisting data into storage
+     */
+    static storageKey = 'trackerStats'
+    /**
+     * The key to use when we don't want to record the company name
+     */
+    static unknownCompanyKey = 'unknown'
+    /**
+     * The prefix used for events. This is used to ensure we only handle events we care about
+     */
+    static eventPrefix = 'newTabPage_'
+
+    /**
+     * @type {Map<string, number> | null}
+     */
+    top100Companies = null
+
+    /**
+     * Internal flag to enable some certain types of logging when developing
+     */
+    _debug = false
+
+    /**
+     * @param {import("./classes/tracker-stats").TrackerStats} stats
+     */
+    constructor (stats) {
+        this.stats = stats
+    }
+
+    /**
+     * Register *all* communications with the extension here.
+     *
+     * The purpose of this method is to co-locate all extension handlers in a single
+     * place for this module.
+     */
+    register () {
+        /**
+         * This listener will redirect the request for tracker-stats.html
+         * on the new tab page to our own HTML file under `web_accessible_resources`
+         */
+        browser.webRequest.onBeforeRequest.addListener(redirectIframeForTrackerStatsMV2,
+            {
+                // todo(Shane): limit this to only the urls we care about
+                urls: ['<all_urls>']
+            },
+            ['blocking']
+        )
+
+        /**
+         * Handle runtime messages sent from the new tab page
+         */
+        browser.runtime.onMessage.addListener((event, sender) => {
+            if (sender.id !== browserWrapper.getExtensionId()) return
+            this.handleIncomingEvent(event)
+        })
```

Comment:
Is there a potential conflict here with the existing central message listener? https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/shared/js/background/events.es6.js#L375

Particularly in the case where a return value is expected there, could this method returning undefined break the flow?

Reply:
I'm moving to Ports to solve this :)